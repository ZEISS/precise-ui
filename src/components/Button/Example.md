**Elementary**

A simple anchor button.

```jsx
const { Button } = require('precise-ui');

<Button href="#🍕">Click me</Button>
```

Custom interaction after clicking the button.

```jsx
const { Button } = require('precise-ui');

<Button onClick={() => alert('Hi')}>Click me</Button>
```

**Button Styles**

There are two types of buttons with their own color themes: `primary` (default) and `secondary`.

```jsx
const { Button } = require('precise-ui');

<div>
  <div>
    <Button>Default</Button>
    <Button buttonStyle="primary">Primary</Button>
    <Button buttonStyle="secondary">Secondary</Button>
  </div>
</div>

```

Button size could be changed with help of `size` property. The following example demonstrates small buttons.

```jsx
const { Button } = require('precise-ui');

  <div>
    <Button size='small' buttonStyle="primary">Primary</Button>
    <Button size='small' buttonStyle="secondary">Secondary</Button>
  </div>
```

With icons on either position.

```jsx
const { Button } = require('precise-ui');

<div>
  <Button href="#🍕" icon="Close" theme={{ buttonIconPosition: 'left'}} buttonStyle='secondary'>Abort</Button>
  <Button href="#🍕" icon="Check">Accept</Button>
  <Button href="#🍕" icon="Check" size="small">Accept small</Button>
</div>
```

Non-responsive Button: it does not change to full-width and remove marings on screens with less than 980px width.

```jsx
const { Button } = require('precise-ui');

  <div>
    <Button disableMobileFullWidth>Non-responsive Button 1</Button>
    <Button disableMobileFullWidth>Non-responsive Button 2</Button>
  </div>
```
