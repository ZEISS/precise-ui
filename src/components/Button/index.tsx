import * as React from 'react';
import styled, { themed, css } from '../../utils/styled';
import { StandardProps, ButtonThemeSettings, PreciseFullTheme } from '../../common';
import { Anchor, AnchorProps } from '../Anchor';
import { IconName, Icon, IconProps } from '../Icon';
import { distance } from '../../distance';
import { remCalc } from '../../utils/remCalc';
import { displayTo } from '../../utils/displayTo';

/**
 * Button style name.
 */
export type ButtonStyle = 'primary' | 'secondary';

/**
 * Button type.
 */
export type ButtonType = 'submit' | 'reset' | 'button';

/**
 * Button size.
 */
export type ButtonSize = 'small' | 'medium';

export interface ButtonProps extends AnchorProps {
  /**
   * Sets the button as active.
   * @default false
   */
  active?: boolean;
  /**
   * Sets the button as a block.
   * @default false
   */
  block?: boolean;
  /**
   * Sets the button style type. By default primary.
   * @default primary
   */
  buttonStyle?: ButtonStyle;
  /**
   * Sets the button type. By default button.
   * @default button
   */
  type?: ButtonType;
  /**
   * Sets the button size. By default medium.
   * @default medium
   */
  size?: ButtonSize;
  /**
   * The name of the icon to display. By default no icon is display.
   */
  icon?: IconName;
}

interface StyledButtonProps extends AnchorProps {
  block?: boolean;
  buttonStyle?: ButtonStyle;
  size?: ButtonSize;
}

interface IconWrapperProps extends StandardProps {}

function getThemeSettings(theme: PreciseFullTheme, buttonStyle?: ButtonStyle) {
  switch (buttonStyle) {
    default:
    case 'primary':
      return theme.buttonPrimary;

    case 'secondary':
      return theme.buttonSecondary;
  }
}

function getButtonLineHeight(themeSettings: ButtonThemeSettings, size?: ButtonSize) {
  switch (size) {
    default:
    case 'medium':
      return themeSettings.lineHeightMedium;

    case 'small':
      return themeSettings.lineHeightSmall;
  }
}

function getButtonFontSize(size?: ButtonSize) {
  switch (size) {
    default:
    case 'medium':
      return remCalc('16px');

    case 'small':
      return remCalc('14px');
  }
}

function getIconSize(size?: ButtonSize) {
  switch (size) {
    default:
    case 'medium':
      return '22px';

    case 'small':
      return '16px';
  }
}

function getButtonPadding(size?: ButtonSize) {
  switch (size) {
    default:
    case 'medium':
      return `9px ${distance.medium} 7px`;

    case 'small':
      return `5px ${distance.small} 3px`;
  }
}

const PseudoButtonStyle = (colorTheme: ButtonThemeSettings) => css`
  &:hover {
    background-color: ${colorTheme.hoverBackground};
    border: ${colorTheme.hoverBorder};
    color: ${colorTheme.hoverText};
  }
  &:focus {
    background-color: ${colorTheme.focusBackground};
    border: ${colorTheme.focusBorder};
    color: ${colorTheme.focusText};
  }
`;

const StyledButton = styled<StyledButtonProps>(
  ({ block: _0 = false, borderless: _1 = false, buttonStyle: _2 = '', size: _3 = '', ...rest }) => <Anchor {...rest} />,
)`
  box-sizing: border-box;
  outline: none;
  border-radius: 0;
  margin: ${distance.small};
  &:first-of-type {
    margin-left: 0;
  }
  &:last-of-type {
    margin-right: 0;
  }
  border: ${themed(props =>
    props.disabled
      ? `${getThemeSettings(props.theme, props.buttonStyle).disabledBorder}`
      : `${getThemeSettings(props.theme, props.buttonStyle).border}`,
  )};
  background-color: ${themed(props =>
    props.disabled
      ? getThemeSettings(props.theme, props.buttonStyle).disabledBackground
      : getThemeSettings(props.theme, props.buttonStyle).background,
  )};
  color: ${themed(props =>
    props.disabled
      ? getThemeSettings(props.theme, props.buttonStyle).disabledText
      : getThemeSettings(props.theme, props.buttonStyle).text,
  )};
  font-family: ${themed(props => props.theme.fontFamily)};
  font-size: ${props => getButtonFontSize(props.size)};
  line-height: ${themed(props => getButtonLineHeight(getThemeSettings(props.theme, props.buttonStyle), props.size))};
  padding: ${props => getButtonPadding(props.size)};
  display: ${props => (props.block ? 'block' : 'inline-block')};
  cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
  ${themed(props => (!props.disabled ? PseudoButtonStyle(getThemeSettings(props.theme, props.buttonStyle)) : ''))};
  ${displayTo('smallAndMedium')`
    width: 100%;
    margin: ${distance.small} 0;
  `};
`;

const DefaultWrapper = styled.div``;

const WithIconWrapper = styled<IconWrapperProps, 'div'>('div')`
  ${props =>
    props.theme.buttonIconPosition === 'left'
      ? `padding-left: ${distance.xlarge}`
      : `padding-right: ${distance.xlarge}`};
  position: relative;
`;

const StyledIcon = styled<IconWrapperProps & IconProps>(Icon)`
  ${props => (props.theme.buttonIconPosition === 'left' ? 'left: 0' : 'right: 0')};
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
`;

/**
 * The button component renders a simple button optionally with an icon.
 */
export const Button: React.SFC<ButtonProps> = ({ children, icon, size, theme, ...rest }) => {
  const Wrapper = icon ? WithIconWrapper : DefaultWrapper;

  return (
    <StyledButton tagName="button" theme={theme} size={size} {...rest}>
      <Wrapper theme={theme}>
        {children}
        {icon && <StyledIcon name={icon} theme={theme} size={getIconSize(size)} />}
      </Wrapper>
    </StyledButton>
  );
};
Button.displayName = 'Button';
