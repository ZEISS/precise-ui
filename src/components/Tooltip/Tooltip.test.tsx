import * as React from 'react';
import * as enzyme from 'enzyme';
import { Tooltip } from './';
import { Button } from '../Button';

describe('<Tooltip />', () => {
  it('should render a tooltip with simple text content over the wrapped component', () => {
    const wrapper = enzyme.shallow(<Tooltip content={'Tooltip content'}>Wrapped element</Tooltip>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render an open tooltip (controlled state)', () => {
    const wrapper = enzyme.shallow(
      <Tooltip open content={'Tooltip content'}>
        Wrapped element
      </Tooltip>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a tooltip with custom position (auto)', () => {
    const wrapper = enzyme.shallow(<Tooltip content={'Tooltip content'}>Wrapped element</Tooltip>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render an open tooltip with custom position (top)', () => {
    const wrapper = enzyme.shallow(
      <Tooltip open position={'top'} content={'Tooltip content'}>
        Wrapped element
      </Tooltip>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render an open tooltip with custom position (left)', () => {
    const wrapper = enzyme.shallow(
      <Tooltip open position={'left'} content={'Tooltip content'}>
        Wrapped element
      </Tooltip>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a tooltip over the complex component', () => {
    const wrapper = enzyme.shallow(
      <Tooltip open content={'Tooltip content'}>
        <Button>Button</Button>
      </Tooltip>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a tooltip over the complex content', () => {
    const wrapper = enzyme.shallow(
      <Tooltip
        content={
          <div style={{ width: '200px' }}>
            <h4 style={{ textAlign: 'center' }}>Header</h4>
            <span>Text content</span>
          </div>
        }>
        Wrapped element
      </Tooltip>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
