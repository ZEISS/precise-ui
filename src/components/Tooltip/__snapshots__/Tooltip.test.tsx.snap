// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> should render a tooltip over the complex component 1`] = `
<styled.div
  innerRef={[Function]}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Flyout
    content="Tooltip content"
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    excludeScrollbar={false}
    open={true}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    preventDefault={false}
    stopPropagation={false}
    theme={
      Object {
        "flyout": Object {
          "background": "rgba(75,78,82,1)",
          "fontSize": "0.875rem",
          "maxWidth": "250px",
          "textColor": "rgba(255,255,255,1)",
        },
      }
    }
  >
    <Button>
      Button
    </Button>
  </Flyout>
</styled.div>
`;

exports[`<Tooltip /> should render a tooltip over the complex content 1`] = `
<styled.div
  innerRef={[Function]}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Flyout
    content={
      <div
        style={
          Object {
            "width": "200px",
          }
        }
      >
        <h4
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          Header
        </h4>
        <span>
          Text content
        </span>
      </div>
    }
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    excludeScrollbar={false}
    open={false}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    preventDefault={false}
    stopPropagation={false}
    theme={
      Object {
        "flyout": Object {
          "background": "rgba(75,78,82,1)",
          "fontSize": "0.875rem",
          "maxWidth": "250px",
          "textColor": "rgba(255,255,255,1)",
        },
      }
    }
  >
    Wrapped element
  </Flyout>
</styled.div>
`;

exports[`<Tooltip /> should render a tooltip with custom position (auto) 1`] = `
<styled.div
  innerRef={[Function]}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Flyout
    content="Tooltip content"
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    excludeScrollbar={false}
    open={false}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    preventDefault={false}
    stopPropagation={false}
    theme={
      Object {
        "flyout": Object {
          "background": "rgba(75,78,82,1)",
          "fontSize": "0.875rem",
          "maxWidth": "250px",
          "textColor": "rgba(255,255,255,1)",
        },
      }
    }
  >
    Wrapped element
  </Flyout>
</styled.div>
`;

exports[`<Tooltip /> should render a tooltip with simple text content over the wrapped component 1`] = `
<styled.div
  innerRef={[Function]}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Flyout
    content="Tooltip content"
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    excludeScrollbar={false}
    open={false}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    preventDefault={false}
    stopPropagation={false}
    theme={
      Object {
        "flyout": Object {
          "background": "rgba(75,78,82,1)",
          "fontSize": "0.875rem",
          "maxWidth": "250px",
          "textColor": "rgba(255,255,255,1)",
        },
      }
    }
  >
    Wrapped element
  </Flyout>
</styled.div>
`;

exports[`<Tooltip /> should render an open tooltip (controlled state) 1`] = `
<styled.div
  innerRef={[Function]}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Flyout
    content="Tooltip content"
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    excludeScrollbar={false}
    open={true}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    preventDefault={false}
    stopPropagation={false}
    theme={
      Object {
        "flyout": Object {
          "background": "rgba(75,78,82,1)",
          "fontSize": "0.875rem",
          "maxWidth": "250px",
          "textColor": "rgba(255,255,255,1)",
        },
      }
    }
  >
    Wrapped element
  </Flyout>
</styled.div>
`;

exports[`<Tooltip /> should render an open tooltip with custom position (left) 1`] = `
<styled.div
  innerRef={[Function]}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Flyout
    content="Tooltip content"
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    excludeScrollbar={false}
    open={true}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    position="left"
    preventDefault={false}
    stopPropagation={false}
    theme={
      Object {
        "flyout": Object {
          "background": "rgba(75,78,82,1)",
          "fontSize": "0.875rem",
          "maxWidth": "250px",
          "textColor": "rgba(255,255,255,1)",
        },
      }
    }
  >
    Wrapped element
  </Flyout>
</styled.div>
`;

exports[`<Tooltip /> should render an open tooltip with custom position (top) 1`] = `
<styled.div
  innerRef={[Function]}
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Flyout
    content="Tooltip content"
    eventTypes={
      Array [
        "mousedown",
        "touchstart",
      ]
    }
    excludeScrollbar={false}
    open={true}
    outsideClickIgnoreClass="ignore-react-onclickoutside"
    position="top"
    preventDefault={false}
    stopPropagation={false}
    theme={
      Object {
        "flyout": Object {
          "background": "rgba(75,78,82,1)",
          "fontSize": "0.875rem",
          "maxWidth": "250px",
          "textColor": "rgba(255,255,255,1)",
        },
      }
    }
  >
    Wrapped element
  </Flyout>
</styled.div>
`;
